src/Microsoft.DotNet.XHarness.Common/Logging/ScanLog.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+#nullable enable
+namespace Microsoft.DotNet.XHarness.Common.Logging
+{
+    /// <summary>
+    /// Log that scans for a given tag and notifies when tag is found in the stream.
+    /// </summary>
+    public class ScanLog : Log
+    {
+        private readonly string _tag;
+        private readonly Action _tagFoundNotify;
+        private readonly char[] _buffer;
+        private int _startIndex;
+        private bool _hasBeenFilled = false;
+
+        public override bool Timestamp { get => false; set { } }
+
+        public ScanLog(string tag, Action tagFoundNotify)
+        {
+            _tag = tag;
+            _tagFoundNotify = tagFoundNotify;
+            _buffer = new char[_tag.Length];
+            _startIndex = -1;
+        }
+
+        protected override void WriteImpl(string value)
+        {
+            foreach (var c in value)
+            {
+                Add(c);
+
+                if (IsMatch())
+                {
+                    _tagFoundNotify();
+                }
+            }
+        }
+
+        public override void Flush()
+        {
+        }
+
+        public override void Dispose()
+        {
+            GC.SuppressFinalize(this);
+        }
+
+        private void Add(char c)
+        {
+            _startIndex++;
+
+            if (_startIndex == _buffer.Length - 1)
+            {
+                _hasBeenFilled = true;
+            }
+
+            _startIndex %= _buffer.Length;
+            _buffer[_startIndex] = c;
+        }
+
+        private bool IsMatch()
+        {
+            if (!_hasBeenFilled)
+            {
+                return false;
+            }
+
+            for (int i = 1; i <= _buffer.Length; i++)
+            {
+                int index = (i + _startIndex) % _buffer.Length;
+                if (_buffer[index] != _tag[i - 1])
```

Comment:
why do you prefer custom search instead of built-in search for strings?

Reply:
It's a round buffer so we go over the end and then from the beginning. I would have to always construct a new string somehow and compare and since all of the logging goes through this I wanted it to be fast.

I didn't find any round buffer in `System.Collections` unfortunately

Reply:
got it, I was asking out of curiosity to learn possible options for collections